### ISSUE TITLE
Suggested way to test if messages are in the  buffer for a receiver?

### ISSUE THREAD

**Issue by ps2goat** [amqp/rhea#109]
I have implemented wrappers around rhea that help to mostly promisify things so that I can await them.  One of the features I added is a `pause` method, so that the receiver stops accepting new messages from the broker, but continues processing everything in the buffer.

Essentially, my method sets the credit window to 0 and returns a promise, which resolves when the `receiver_drained` event occurs.  

However, this event doesn't seem to be firing when there are no messages in the buffer (the queue is empty).  What is your preferred approach to doing this?  I don't want to just close the link, as it will take _n_ seconds before the messages are placed back into the queue (server-side configuration).  I have the receiver configured so that I must manually ACK the server after I'm done processing the messages, so no messages will be lost, only delayed.

Most of the examples I've seen close the receiver after _n_ number of messages have been received.

**Comment by grs** [comment]
Setting the credit window to 0 on an existing receiver will not have any effect at present. There is no way (as yet) to disable an automatically managed credit window once it has been enabled (which it is by default); all you can do is change the window to some other non-zero value.

What you want is the drain behaviour. There is an example for this: https://github.com/amqp/rhea/blob/master/examples/drain.js however it does not use the automatic credit window at all (it is disabled from the start). Would that approach work for you in the short term?

**Comment by ps2goat** [comment]
I am using the drain feature, but incorrectly assumed the credit window behavior.  The only issue I have with the drain feature is that I am monitoring for the `receiver_drained` event, but it seems to only emit if there are already messages in the buffer. If I have zero messages in the buffer, the event does not emit.  That's what I am trying to work around, currently.  I was hoping to not have to inspect the buffer's contents directly.

If I set the receiver to drain and there are zero messages, can/do you emit the `receiver_drained` event?

Aside from that, if I want to continue processing again, do I just set `receiver.drain = false;` ?

**Comment by grs** [comment]
No, you should get the receiver_drained regardless, of any buffered messages, provided the credit is non-zero at the point you set drain=true.

After draining you will need to re-issue credit.
